docs/guides/src/main/server/fips.adoc

```diff
@@ -0,0 +1,229 @@
+
+== BouncyCastle library
+
+Keycloak internally uses the BouncyCastle library for many cryptography utilities. However, the default flavor of the BouncyCastle library that shipped with Keycloak is not FIPS compliant, 
+but, BouncyCastle also provides a FIPS validated version of it's library. The FIPS validated BouncyCastle library cannot be shipped with Keycloak due to license constraints and
+Keycloak cannot provide official support of it. Therefore, to run in FIPS compliant mode, you need to download BouncyCastle-FIPS bits and add them to the Keycloak distribution.
+When Keycloak executes in fips-mode, it will use the BCFIPS bits instead of the default BouncyCastle bits, which achieves FIPS compliance.
+
+=== BouncyCastle FIPS bits
+
+BouncyCastle FIPS can be downloaded from the https://www.bouncycastle.org/fips-java/[BouncyCastle official page]. Then you can add them to the directory
+`KEYCLOAK_HOME/providers` of your distribution. Make sure to use proper versions compatible with BouncyCastle Keycloak dependencies. The supported BCFIPS bits needed are:
+
+* `bc-fips-1.0.2.3.jar`
+* `bctls-fips-1.0.14.jar`
+* `bcpkix-fips-1.0.7.jar`
+
+== Generating keystore
+
+You can create either `pkcs12` or `bcfks` keystore to be used for the Keycloak server SSL. The `pkcs12` works well in BCFIPS non-approved mode.
+
+PKCS12 keystore can be generated with OpenJDK 17 Java on RHEL 9 in the standard way.
+
+BCFKS keystore generation requires the use of the BouncyCastle FIPS libraries and a custom security file.
+
+You can start with create a helper file, such as `/tmp/kc.keystore-create.java.security`. The content of the file needs only to have the following property:
+```
+securerandom.strongAlgorithms=PKCS11:SunPKCS11-NSS-FIPS
+```
+
+Next enter a command such as the following to generate the keystore:
+```
+keytool -keystore $KEYCLOAK_HOME/conf/server.keystore \
+  -storetype bcfks \
+  -providername BCFIPS \
+  -providerclass org.bouncycastle.jcajce.provider.BouncyCastleFipsProvider \
+  -provider org.bouncycastle.jcajce.provider.BouncyCastleFipsProvider \
+  -providerpath $KEYCLOAK_HOME/providers/bc-fips-*.jar \
+  -alias localhost \
+  -genkeypair -sigalg SHA512withRSA -keyalg RSA -storepass passwordpassword \
+  -dname CN=localhost -keypass passwordpassword \
+  -J-Djava.security.properties=/tmp/kc.keystore-create.java.security
+```
+
+WARNING: Using self-signed certificates is for demonstration purposes only, so replace these certificates with proper certificates when you move to a production environment.
+
+Similar options are needed when you are doing any other manipulation with keystore/truststore of `bcfks` type.
+
+== Running the server.
+
+To run the server with BCFIPS in non-approved mode, enter the following command::
+
+<@kc.start parameters="--fips-mode=enabled --hostname=localhost --https-key-store-password=passwordpassword --log-level=INFO,org.keycloak.common.crypto:TRACE,org.keycloak.crypto:TRACE"/>
```

Comment:
Now it can be `--features=fips` only and it defaults to `--fips-mode=non-strict` although the latter option can also be passed with `non-strict` or `strict` values.